Easydict/NewApp/Utility/Extensions/QueryService+ConfigurableService/CustomOpenAIService+ConfigurableService.swift

```diff
@@ -0,0 +1,155 @@
+        )
+        // model
+        TextField(
+            "service.configuration.custom_openai.allmodels.title",
+            text: viewModel.$availableModels,
+            prompt: Text("service.configuration.custom_openai.model.placeholder")
+        )
+        .padding(10.0)
+        .lineLimit(1 ... 3)
+        Picker(
+            "service.configuration.openai.model.title",
+            selection: viewModel.$model
+        ) {
+            ForEach(viewModel.validModels, id: \.self) { value in
+                Text(value)
+            }
+        }
+        .padding(10.0)
+
+        ServiceConfigurationToggleCell(
+            titleKey: "service.configuration.openai.translation.title",
+            key: .customOpenAITranslation
+        )
+        ServiceConfigurationToggleCell(
+            titleKey: "service.configuration.openai.sentence.title",
+            key: .customOpenAISentence
+        )
+        ServiceConfigurationToggleCell(
+            titleKey: "service.configuration.openai.dictionary.title",
+            key: .customOpenAIDictionary
+        )
+        ServiceConfigurationPickerCell(
+            titleKey: "service.configuration.openai.usage_status.title",
+            key: .customOpenAIServiceUsageStatus,
+            values: OpenAIUsageStats.allCases
+        )
+    }
+
+    // MARK: Private
+
+    @ObservedObject private var viewModel: CustomOpenAIViewModel
+}
+
+// MARK: - CustomOpenAIViewModel
+
+private class CustomOpenAIViewModel: ObservableObject {
+    // MARK: Lifecycle
+
+    init(service: CustomOpenAIService) {
+        self.service = service
+        cancellables.append(
+            Defaults.publisher(keys: [.customOpenAINameKey, .customOpenAIModel])
+                .sink { _ in
+                    self.serviceConfigChanged()
+                }
+        )
+        cancellables.append(
+            Defaults.publisher(keys: [.customOpenAIModelsAvailable])
+                .sink { _ in
+                    self.modelsTextChanged()
+                }
+        )
+    }
+
+    // MARK: Internal
+
+    let service: CustomOpenAIService
+
+    @Default(.customOpenAIModel) var model
+    @Default(.customOpenAIModelsAvailable) var availableModels
+
+    @Published var validModels: [String] = []
+
+    // MARK: Private
+
+    private var cancellables: [AnyCancellable] = []
+
+    private func modelsTextChanged() {
+        if availableModels.isEmpty {
+            model = ""
+            validModels = []
+            return
+        }
+        validModels = availableModels.components(separatedBy: ",").filter { !$0.isEmpty }
+        if validModels.count == 1 || !validModels.contains(model) {
+            model = validModels[0]
+        }
+    }
+
+    private func serviceConfigChanged() {
+        let userInfo: [String: Any] = [EZWindowTypeKey: service.windowType.rawValue]
+        let notification = Notification(name: .serviceHasUpdated, object: nil, userInfo: userInfo)
```

Comment:
```suggestion
        let notification = Notification(name: .serviceHasUpdated, object: nil, userInfo: nil)
```

All windows currently share these services, so when you modify the service configuration, you need to update all window service information.

When userInfo is nil, it means to send update notification to all windows.

Reply:
I think it's ok to pass the window type based on my test. The main window is updated while I edit the service info in the mini window segment.